File: src/main/kotlin/easy/ValidParentheses.kt
package easy

object ValidParentheses {
    /*
   * Link: https://leetcode.com/problems/valid-parentheses/
   * Difficulty: Easy
   */

    private val charMap = mapOf(
            Pair('{', 1),
            Pair('}', -1),
            Pair('[', 2),
            Pair(']', -2),
            Pair('(', 3),
            Pair(')', -3)
    )


    fun isValid(s: String): Boolean {
        val stack = ArrayDeque<Int>()

        for(c in s) {
            val number = charMap[c]!!
            if(number > 0) {
                // Quando é um sinal de abrir
                stack.addLast(number)
            } else {
                if(stack.isEmpty()) return false
                // Quando é um sinal de fechar
                if (stack.last() + number == 0) {
                    stack.removeLast()
                } else return false
            }
        }
        return stack.isEmpty()
    }
}

File: src/easy/valid_parentheses.go
package easy

/*
Difficult: Easy
Link: https://leetcode.com/problems/valid-parentheses/
*/
func isValid(s string) bool {
	var stack []rune
	mapping := map[rune]rune{
		'}': '{',
		']': '[',
		')': '(',
	}
	// Time: O(N)
	// Space: O(N)
	for _, c := range s {
		n := len(stack) - 1
		if value, ok := mapping[c]; ok && n >= 0 {
			if value != stack[n] {
				return false
			}
			stack = stack[:n]
		} else {
			stack = append(stack, c)
		}
	}
	return len(stack) == 0
}
